triggers/daml/Daml/Trigger/LowLevel.daml

```diff
@@ -244,3 +255,111 @@ data Commands = Commands
   { commandId : CommandId
   , commands : [Command]
   }
+
+-- Same as in Script; TODO factor
+data Free f a
+  = Pure a
+  | Free (f (Free f a))
+
+instance Functor f => Functor (Free f) where
+  fmap f (Pure a) = Pure (f a)
+  fmap f (Free x) = Free (fmap f <$> x)
+
+instance Functor f => Applicative (Free f) where
+  pure = Pure
+  Pure f <*> Pure a = Pure (f a)
+  Pure f <*> Free x = Free (fmap f <$> x)
+  Free x <*> my     = Free ((<*> my) <$> x)
+
+instance Functor f => Action (Free f) where
+  Pure a >>= f = f a
+  Free x >>= f = Free ((>>= f) <$> x)
+
+liftFC : Functor f => f a -> Free f a
+liftFC fa = Free (Pure <$> fa)
+
+lowerFC : Action f => Free f a -> f a
+lowerFC (Pure a) = pure a
+lowerFC (Free x) = x >>= lowerFC
+
+newtype StateT s m a = StateT { runStateT : s -> m (a, s) }
+  deriving Functor
+
+liftStateT : Functor m => m a -> StateT s m a
+liftStateT ma = StateT $ \s -> (,s) <$> ma
+
+instance Action m => Applicative (StateT s m) where
+  pure a = StateT (\s -> pure (a, s))
+  f <*> x = f >>= (<$> x)
+  
+instance Action m => Action (StateT s m) where
+  StateT x >>= f = StateT $ \s -> do
+    (x', s') <- x s
+    runStateT (f x') s'
+
+class ActionState s m | m -> s where
+  get : m s
+  put : s -> m ()
+
+modify : (Action m, ActionState s m) => (s -> s) -> m ()
+modify f = put . f =<< get
+
+instance Applicative m => ActionState s (StateT s m) where
+  get = StateT $ \s -> pure (s, s)
+  put s = StateT $ const $ pure ((), s)
+
+execStateT : Functor m => StateT s m a -> s -> m s
+execStateT (StateT fa) = fmap snd . fa
+
+zoom : Functor m => (t -> s) -> (t -> s -> t) -> StateT s m a -> StateT t m a
+zoom r w (StateT smas) = StateT $ \t ->
+  smas (r t) <&> \(a, s) -> (a, w t s)
+
+-- Must be kept in sync with Runner#handleStepFreeResult
+data TriggerF a =
+  GetTime (Time -> a)
+  | Submit (Commands, () -> a)
```

Comment:
As mentioned on the issue, one thing that we can do with this monadified version is to return a `CommandId` here so we don’t have to do this annoying back and forth translation between the random UUIDs in the trigger runner and what users specify here. Let’s leave that for a separate PR though.